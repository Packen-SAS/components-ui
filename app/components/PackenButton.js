import React, { Component } from "react";
import { TouchableWithoutFeedback, View } from "react-native"

import Icon from "react-native-vector-icons/Feather";
import Color from "../styles/abstracts/colors";
import PackenText from "./PackenText";
import * as ButtonStyles from "../styles/components/PackenButton";

class PackenButton extends Component {
  constructor(props) {
    super(props);

    styles = {
      shape: {
        ...ButtonStyles.base.shape,
        ...ButtonStyles[props.type][props.size].shape,
        ...ButtonStyles[props.level].shape
      },
      content: {
        position: "relative"
      },
      icon: {
        ...ButtonStyles.base.icon,
        ...ButtonStyles[props.type][props.size].icon,
        ...ButtonStyles[props.level].icon
      }
    };

    switch (props.type) {
      case "icon":
        styles = {
          ...styles,
          iconWrapper: {
            position: "absolute",
            top: 0,
            left: 0
          }
        };
        break;
      case "regular":
        styles = {
          ...styles,
          label: {
            ...ButtonStyles.base.label,
            ...ButtonStyles[props.type][props.size].label,
            ...ButtonStyles[props.level].label
          },
          iconWrapper: {
            position: "absolute",
            top: 0,
            right: props.icon ? props.icon.position === "left" ? "auto" : -(ButtonStyles[props.type][props.size].label.marginHorizontal + 0) : 0,
            left: props.icon ? props.icon.position === "right" ? "auto" : -(ButtonStyles[props.type][props.size].label.marginHorizontal + 0) : 0
          }
        };
        break;
      default:
        break;
    }

    switch (props.level) {
      case "secondary":
        styles.shape = {
          ...styles.shape,
          borderWidth: 1,
          borderStyle: "solid",
          borderColor: Color.secondary.default_drk
        }
        break;
      default:
      case "ghost":
      case "primary":
      case "tertiary":
      case "danger":
        break;
    }

    if (props.isDisabled) {
      styles.shape.backgroundColor = Color.base.disabled;
      styles.icon.color = Color.base.white;

      if (styles.label) {
        styles.label.color = Color.base.white;
      }

      if (styles.shape.borderWidth) {
        styles.shape.borderWidth = 0;
      }

      if (props.level === "ghost") {
        styles.shape.backgroundColor = Color.base.transparent;
        styles.icon.color = Color.base.disabled_alt;

        if (styles.label) {
          styles.label.color = Color.base.disabled_alt;
        }
      }
    }

    this.state = {
      type: props.type,
      level: props.level,
      size: props.size,
      icon: props.icon ? props.icon : undefined,
      isDisabled: props.isDisabled,
      shapeHeight: 0,
      shapeWidth: 0,
      iconHeight: 0,
      iconWidth: 0,
      styles: styles
    }
  }

  get_styles = () => {
    let styles = {
      shape: {
        ...ButtonStyles.base.shape,
        ...ButtonStyles[this.props.type][this.props.size].shape,
        ...ButtonStyles[this.props.level].shape
      },
      content: {
        position: "relative"
      },
      icon: {
        ...ButtonStyles.base.icon,
        ...ButtonStyles[this.props.type][this.props.size].icon,
        ...ButtonStyles[this.props.level].icon
      }
    };

    switch (this.props.type) {
      case "icon":
        styles = {
          ...styles,
          iconWrapper: {
            position: "absolute",
            top: (this.state.shapeHeight/2) - (this.state.iconHeight/2),
            left: (this.state.shapeWidth/2) - (this.state.iconWidth/2)
          }
        };
        break;
      case "regular":
        styles = {
          ...styles,
          label: {
            ...ButtonStyles.base.label,
            ...ButtonStyles[this.props.type][this.props.size].label,
            ...ButtonStyles[this.props.level].label
          },
          iconWrapper: {
            position: "absolute",
            top: (this.state.shapeHeight/2) - (this.state.iconHeight/2),
            right: this.props.icon ? this.props.icon.position === "left" ? "auto" : -(ButtonStyles[this.props.type][this.props.size].label.marginHorizontal + (this.state.iconWidth/2)) : 0,
            left: this.props.icon ? this.props.icon.position === "right" ? "auto" : -(ButtonStyles[this.props.type][this.props.size].label.marginHorizontal + (this.state.iconWidth/2)) : 0
          }
        };
        break;
      default:
        break;
    }

    /* Custom styles depending on the level */
    switch (this.props.level) {
      case "secondary":
        styles.shape = {
          ...styles.shape,
          borderWidth: 1,
          borderStyle: "solid",
          borderColor: Color.secondary.default_drk
        }
        break;
      default:
      case "ghost":
      case "primary":
      case "tertiary":
      case "danger":
        break;
    }

    /* Custom styles for isDisabled states */
    if (this.props.isDisabled) {
      styles.shape.backgroundColor = Color.base.disabled;
      styles.icon.color = Color.base.white;

      if (styles.label) {
        styles.label.color = Color.base.white;
      }

      if (styles.shape.borderWidth) {
        styles.shape.borderWidth = 0;
      }

      if (this.props.level === "ghost") {
        styles.shape.backgroundColor = Color.base.transparent;
        styles.icon.color = Color.base.disabled_alt;

        if (styles.label) {
          styles.label.color = Color.base.disabled_alt;
        }
      }
    }

    return styles;
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevState.shapeHeight !== this.state.shapeHeight || prevState.iconHeight !== this.state.iconHeight || prevState.iconWidth !== this.state.iconWidth) {
      this.setState({
        styles: this.get_styles()
      });
    }
  }

  get_shape_dimensions = e => {
    this.setState({
      shapeHeight: Math.floor(e.height),
      shapeWidth: Math.floor(e.width)
    });
  }

  get_icon_dimensions = e => {
    this.setState({
      iconHeight: Math.floor(e.height),
      iconWidth: Math.floor(e.width)
    });
  }

  execute_callback = () => {
    this.props.callback();
  }

  pressIn_handler = () => {
    let newStyles = {...this.get_styles()}
    newStyles.shape.backgroundColor = Color[this.state.level].focus;

    /* Custom focus styles */
    switch (this.state.level) {
      case "secondary":
        newStyles.shape.borderColor = Color.secondary.focus;
        break;
      default:
      case "primary":
      case "tertiary":
      case "ghost":
      case "danger":
        break;
    }

    this.setState({
      styles: newStyles
    });
  }

  pressOut_handler = () => {
    const newStyles = {...this.get_styles()}
    newStyles.shape.backgroundColor = Color[this.state.level].default;
    this.setState({
      styles: newStyles
    });
  }

  render() {
    return (
      <View pointerEvents={this.state.isDisabled ? "none" : "auto"}>
        <TouchableWithoutFeedback onPress={this.execute_callback} onPressIn={this.pressIn_handler} onPressOut={this.pressOut_handler}>
          <View style={this.state.styles.shape}>
            <View style={this.state.styles.shape__content} onLayout={e => { this.get_shape_dimensions(e.nativeEvent.layout); }}>
              <PackenText style={this.state.styles.label}>{this.props.children}</PackenText>
              {
                this.props.icon ? (
                  <View style={this.state.styles.iconWrapper} onLayout={e => { this.get_icon_dimensions(e.nativeEvent.layout); }}>
                    <Icon name={this.props.icon.name} size={this.state.styles.icon.fontSize} color={this.state.styles.icon.color}/>
                  </View>
                ) : null
              }
            </View>
          </View>
        </TouchableWithoutFeedback>
      </View>
    );
  }
}

export default PackenButton;