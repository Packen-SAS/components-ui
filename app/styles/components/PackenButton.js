import Color from "../abstracts/colors";

const base = {
  textAlign: "center",
  alignItems: "center",
  justifyContent: "center"
}

export const icon = {
  tiny: {
    height: tiny.minHeight,
    width: tiny.minHeight
  }
}

export const regular = {
  tiny: {
    ...tiny,
    paddingVertical: 7,
    paddingHorizontal: 11
  }
}

export const primary = {
  ...base,
  bg: {
    default: Color.primary.default,
    focus: Color.primary.focus
  }
}

export const tiny = {
  minHeight: 24
}

export const small = {
  minHeight: 32
}

export const medium = {
  minHeight: 40
}

export const larger = {
  minHeight: 48
}

export const giant = {
  minHeight: 56
}