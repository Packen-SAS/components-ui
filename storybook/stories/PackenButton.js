import React from "react";
import { storiesOf } from "@storybook/react-native";

import Wrapper from "../wrapper";
import PackenButton from "../../app/components/PackenButton";

storiesOf("PackenButton", module)
  .add("Icon primary tiny", () => (
    <Wrapper>
      <PackenButton
        icon={{ name: "arrow-right-circle" }}
        type="icon"
        level="primary"
        size="tiny"
        callback={() => { return; }} />
    </Wrapper>
  ))
  .add("Icon primary small", () => (
    <Wrapper>
      <PackenButton
        icon={{ name: "arrow-right-circle" }}
        type="icon"
        level="primary"
        size="small"
        callback={() => { return; }} />
    </Wrapper>
  ))
  .add("Icon primary medium", () => (
    <Wrapper>
      <PackenButton
        icon={{ name: "arrow-right-circle" }}
        type="icon"
        level="primary"
        size="medium"
        callback={() => { return; }} />
    </Wrapper>
  ))
  .add("Icon primary large", () => (
    <Wrapper>
      <PackenButton
        icon={{ name: "arrow-right-circle" }}
        type="icon"
        level="primary"
        size="large"
        callback={() => { return; }} />
    </Wrapper>
  ))
  .add("Icon primary large disabled", () => (
    <Wrapper>
      <PackenButton
        icon={{ name: "arrow-right-circle" }}
        type="icon"
        level="primary"
        size="giant"
        callback={() => { return; }}
        isDisabled />
    </Wrapper>
  ));