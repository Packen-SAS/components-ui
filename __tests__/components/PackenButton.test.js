import "react-native";
import React from "react";

import PackenButton from "../../app/components/PackenButton";

import renderer from "react-test-renderer";

describe("<PackenButton/>", () => {
  describe("Rendering", () => {
    it("renders a regular button correctly", () => {
      const regularRender = renderer.create(
        <PackenButton
          icon={{ name: "arrow-right", position: "right" }}
          type="regular"
          level="primary"
          size="medium"
          callback={jest.fn}>Medium</PackenButton>
      ).toJSON();

      expect(regularRender).toMatchSnapshot();
    });

    it("renders an icon button correctly", () => {
      const iconRender = renderer.create(
        <PackenButton
          icon={{ name: "arrow-right-circle" }}
          type="icon"
          level="secondary"
          size="large"
          callback={jest.fn} />
      ).toJSON();
      expect(iconRender).toMatchSnapshot();
    });
  });

  describe("styles", () => {
    const render = renderer.create(
      <PackenButton
        icon={{ name: "arrow-right", position: "right" }}
        type="regular"
        level="primary"
        size="medium"
        callback={jest.fn}>Medium</PackenButton>
    );
    const instance = render.getInstance();

    it("returns current styles", () => {
      expect(instance.get_styles()).toBeDefined();
    });
  });
});