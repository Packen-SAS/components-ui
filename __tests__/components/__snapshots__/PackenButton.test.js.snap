// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackenButton/> Rendering renders a regular button correctly 1`] = `
<View
  pointerEvents="auto"
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#00B3DB",
        "flexDirection": "row",
        "justifyContent": "center",
        "minHeight": 40,
        "paddingHorizontal": 46,
        "paddingVertical": 12,
        "position": "relative",
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Muli-Bold",
            "fontSize": 14,
            "letterSpacing": 1,
            "lineHeight": 16,
            "marginHorizontal": 12,
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Medium
      </Text>
      <View
        onLayout={[Function]}
        style={
          Object {
            "left": "auto",
            "position": "absolute",
            "right": -12,
            "top": 0,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 21,
              },
              undefined,
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<PackenButton/> Rendering renders an icon button correctly 1`] = `
<View
  pointerEvents="auto"
>
  <View
    accessible={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "backgroundColor": "#F5F7F8",
        "borderColor": "#556D87",
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
        "position": "relative",
        "width": 48,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <Text
        style={
          Object {
            "color": "#222B45",
            "fontFamily": "Muli-Regular",
            "fontSize": 14,
          }
        }
      />
      <View
        onLayout={[Function]}
        style={
          Object {
            "left": 0,
            "position": "absolute",
            "top": 0,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#556D87",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "Feather",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
